odbc/src/api.rs

```diff
@@ -431,8 +473,54 @@ pub extern "C" fn SQLForeignKeysW(
 }
 
 #[no_mangle]
-pub extern "C" fn SQLFreeHandle(_handle_type: HandleType, _handle: Handle) -> SqlReturn {
-    unimplemented!()
+pub extern "C" fn SQLFreeHandle(handle_type: HandleType, handle: Handle) -> SqlReturn {
+    unsafe {
+        match handle_type {
+            // By making Boxes to the types and letting them go out of
+            // scope, they will be dropped.
+            HandleType::Env => {
+                let env = Box::from_raw(handle as *mut EnvHandle);
+                // Actually reading this value would make ASAN fail, but this
+                // is what the ODBC standard expects.
+                env.env.write().unwrap().state = EnvState::Unallocated;
+            }
+            HandleType::Dbc => {
+                let handle = handle as *mut ConnectionHandle;
+                let conn = Box::from_raw(handle);
+                let mut contents = conn.connection.write().unwrap();
+                // Actually reading this value would make ASAN fail, but this
+                // is what the ODBC standard expects.
+                contents.state = ConnectionState::AllocatedEnvUnallocatedConnection;
+                (*contents.env)
+                    .env
+                    .write()
+                    .unwrap()
+                    .connections
+                    .remove(&handle);
+            }
+            HandleType::Stmt => {
+                let handle = handle as *mut StatementHandle;
+                let stmt = Box::from_raw(handle);
+                let mut contents = stmt.stmt.lock().unwrap();
+                // Actually reading this value would make ASAN fail, but this
+                // is what the ODBC standard expects.
+                contents.state = StatementState::Unallocated;
+                (*contents.connection)
+                    .connection
+                    .write()
+                    .unwrap()
+                    .statements
+                    .remove(&handle);
+            }
+            HandleType::Desc => {
+                let desc = Box::from_raw(handle as *mut DescriptorHandle);
+                // Actually reading this value would make ASAN fail, but this
+                // is what the ODBC standard expects.
```

Comment:
can you elaborate? I don't really know what this comment means.

Reply:
A program reading this state value after `SQLFreeHandle` is reading unallocated memory. ASAN will fail if that happens.

Reply:
I'm arguing that I'm writing a value that should never be read, but it is what is implied by the ODBC standard